### PR TITLE
feat(tracker): Gantt — #2 read-only viewlet

### PR DESCRIPTION
**Pull request by MichaelUray**
## Summary

Read-only Gantt **viewlet** for Tracker: a horizontally-scrolling time-canvas with a synced sidebar that lists the issues in the current project / filter, plus a zoom toolbar (Day / Week / Month / Quarter), a "Today" jump, page-scroll buttons, and customizable sidebar columns. No editing yet — drag/resize comes in PR3a, dependencies in PR3b.

The viewlet registers **after** Kanban so List stays the project default. All Gantt logic lives behind a `viewlet.IssueGantt` registration; no existing viewlet is mutated.

> **Supersedes #10852** (closed). The original head SHA was `7b197a93f`; the branch has since been rebased onto the current PR1 tip and DCO-cleaned.

## What's in this PR (+33 incremental commits)

- `feat(tracker)` register `viewlet.IssueGantt` (placeholder) → wired with reactive queries
- Gantt **library** modules — `time-scale`, `layout`, `types`, types-tested with Jest
- Svelte components — `GanttCanvas`, `GanttHeader`, `GanttSidebar`, `GanttBar`, `GanttTodayMarker`, `GanttMilestoneFlag`, `GanttToolbar`
- `feat(tracker)` register `startDate` + `dueDate` as **Issue filters**
- Hardened scrollbars (sticky-bottom horizontal scrollbar, Plane-style)
- Status-badge column + `ganttShowStatus` toggle
- Inline-add row in the Gantt sidebar
- "+ New issue" button in the Gantt toolbar (wired into the standard NewIssue flow)
- i18n keys for the new Gantt UI strings (en + de)

## Stack overview

| # | Branch | Scope | Status |
|---|---|---|---|
| [PR1](https://github.com/hcengineering/platform/pull/10851) | `feat/gantt-upstream-pr1-schema` | schema + version bump | merged ✅ |
| **PR2** | `feat/gantt-upstream-pr2-readonly` | **this PR** — read-only viewlet | draft |
| [PR3a](https://github.com/hcengineering/platform/pull/10854) | `feat/gantt-upstream-pr3a-edit` | editable viewlet | draft |
| [PR3b](https://github.com/hcengineering/platform/pull/10855) | `feat/gantt-upstream-pr3b-deps-cascade-cp` | dependencies + cascade + critical path | draft |
| [PR4](https://github.com/hcengineering/platform/pull/10856) | `feat/gantt-upstream-pr4-polish` | polish | draft |
| [PR5](https://github.com/hcengineering/platform/pull/10857) | `feat/gantt-upstream-pr5-tier2` | undo, saved views, bulk, auto-scheduling | draft |
| [PR6](https://github.com/hcengineering/platform/pull/10858) | `feat/gantt-upstream-pr6-tier3-virtualization` | virtualization | draft |
| [PR7](https://github.com/hcengineering/platform/pull/10859) | `feat/gantt-upstream-pr7-tier4` | mobile + tree + predecessor col + visual deps + notifications | draft |
| [Docs (huly-docs#70)](https://github.com/hcengineering/huly-docs/pull/70) | `MichaelUray:feat/gantt-section` | reference + screenshots | open |

Marked as **draft** to signal stack dependency on PR1. Will be promoted to ready-for-review once PR1 is merged.

## How to review

GitHub shows the **cumulative diff** against `develop` (PR1 + PR2 = 40 commits, ~2 850 line additions). For the **PR2-only incremental diff** see:

→ [Fork compare: PR1...PR2 — 33 commits, 34 files](https://github.com/MichaelUray/huly-platform/compare/feat/gantt-upstream-pr1-schema...feat/gantt-upstream-pr2-readonly)

## Testing

- Jest tests on the `time-scale` and `layout` modules cover the date-math and row-packing edge cases.
- Validated on a production-style deployment on 2026-05-18.

## DCO

All commits are `Signed-off-by`.

### CODE REVIEW

**Comment by MichaelUray** [comment]
Thanks for merging #10851! I've rebased this PR onto the new `develop` — it now contains only the PR-2 commits (read-only Gantt viewlet); the schema commits dropped out during the rebase as already applied. On our fork CI mirror of this exact head, `build`, `svelte-check`, unit `test`, `uitest`, `uitest-workspaces` and `uitest-qms` are all green; the only red is an intermittent `uitest-pg` failure in an unrelated login sanity test (`LoginPage.checkIfErrorMessageIsShown`) that passes on re-run. The rest of the series (#10854–#10859, #10872, #10873) has been rebased the same way and stays in draft until its predecessor lands. Ready for review.

File: plugins/tracker-resources/src/components/issues/IssuesView.svelte [hunk 1]
```diff
@@ -3,12 +3,16 @@
...
   import { Issue, TrackerEvents } from '@hcengineering/tracker'
-  import { IModeSelector, themeStore } from '@hcengineering/ui'
+  import { Button, IconAdd, IModeSelector, showPopup, themeStore } from '@hcengineering/ui'
   import { ViewOptions, Viewlet } from '@hcengineering/view'
   import { FilterBar, SpaceHeader, ViewletContentView, ViewletSettingButton } from '@hcengineering/view-resources'
   import tracker from '../../plugin'
   import CreateIssue from '../CreateIssue.svelte'
 
+  function newIssue (): void {
+    showPopup(CreateIssue, { space, shouldSaveDraft: true }, 'top')
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/issues/IssuesView.svelte:13]: Should we check that space is defined?

File: plugins/tracker-resources/src/components/gantt/GanttTodayMarker.svelte [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  export let canvasHeight: number
+  // Viewport accepted for API compatibility but no longer used for clipping
+  // (the SVG itself spans the whole canvas so the browser handles overflow).
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  export let viewport: { left: number, right: number } = { left: 0, right: 0 }
+  $: void viewport
+
+  $: today = Date.now()
+  $: x = timeScale.toX(today)
+  $: dateLabel = new Date(today).toLocaleDateString(undefined, { day: 'numeric', month: 'short' })
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttTodayMarker.svelte]: In other places dates snap to UTC midnight, but here `toLocaleDateString` is used.
Was this done intentionally?

File: models/tracker/src/viewlets.ts [hunk 3]
```diff
@@ -212,6 +212,56 @@ export function issueConfig (
   ]
 }
 
+export function ganttViewOptions (): ViewOptionsModel {
+  // PR 2 ships a minimal read-only Gantt. Group-by + Show-colors are
```
Comment (ArtyomSavchenko) [inline models/tracker/src/viewlets.ts]: [minor]Should these comments be cleaned up?
It’s unclear why PR 2 is mentioned in the codebase.

File: plugins/tracker-resources/src/components/gantt/lib/time-scale.ts [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  pxPerDay: number
+  /** Convert a Timestamp to its X coordinate (relative to origin). */
+  toX: (t: number) => number
+  /** Convert an X coordinate back to a snapped Timestamp. */
+  fromX: (x: number) => number
+  /** Generate header ticks across [from, to]. */
+  ticks: (range: [number, number]) => Tick[]
+}
+
+export function createTimeScale (zoom: ZoomLevel, origin: number): TimeScale {
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/lib/time-scale.ts:34]: So far, I only see the function being used in tests.
Will it be used further in other PRs?

File: plugins/tracker-resources/src/components/gantt/GanttBar.svelte [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+      case 'task:statusCategory:UnStarted':
+      case 'tracker:statusCategory:Backlog':
+        return {
+          fill: 'var(--theme-button-default)',
+          border: 'var(--theme-button-border)',
+          text: 'var(--theme-content-color)'
+        }
+      case 'task:statusCategory:ToDo':
+        return { fill: '#dbeafe', border: '#3b82f6', text: '#1e3a8a' }
+      case 'task:statusCategory:Active':
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttBar.svelte:33]: Can we make this more flexible?
I don't like these hardcoded values.

File: plugins/tracker-assets/lang/de.json [hunk 6]
```diff
@@ -293,7 +293,26 @@
...
+    "AllowCreatingIssues": "Erstellen von Aufgaben erlauben",
+    "Day": "Day",
+    "Week": "Week",
+    "Month": "Month",
+    "Quarter": "Quarter",
+    "Gantt": "Gantt",
+    "GanttShowIssueCode": "Show issue code",
+    "GanttShowTitle": "Show title",
+    "GanttShowStatus": "Show status",
+    "GanttToday": "Today",
```
Comment (ArtyomSavchenko) [inline plugins/tracker-assets/lang/de.json]: English translation is used instead of other languages in all locales.

File: plugins/tracker-resources/src/components/gantt/GanttCanvas.svelte [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+    {/each}
+  </g>
+
+  <g class="bars" transform="translate(0, {milestoneStripHeight})">
+    {#each visibleRows as row (rowKey(row))}
+      <!-- svelte-ignore a11y-no-static-element-interactions -->
+      <g
+        class="row-hit"
+        on:mouseenter={(e) => dispatch('hoverRow', { id: row.id, row, mouseX: e.clientX, mouseY: e.clientY })}
+        on:mousemove={(e) => dispatch('hoverRow', { id: row.id, row, mouseX: e.clientX, mouseY: e.clientY })}
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttCanvas.svelte]: hoverRow dispatches on every mousemove and updates tooltipState, re-triggering GanttView reactivity per pixel of mouse travel. Consider only updating when the row id changes, or throttling position updates.

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte [hunk 8]
```diff
@@ -0,0 +1,1153 @@
...
+
+  $: issueDocQuery = (
+    space !== undefined ? { space, ...(query as DocumentQuery<Issue>) } : { ...(query as DocumentQuery<Issue>) }
+  ) as DocumentQuery<Issue>
+  $: milestoneDocQuery = (space !== undefined ? { space } : {}) as DocumentQuery<Milestone>
+  $: issueQuery.query(
+    tracker.class.Issue,
+    issueDocQuery,
+    (res: Issue[]) => {
+      issues = res
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttView.svelte:125]: The issue query has no limit and no virtualization at the query level — every issue in the project/filter is loaded and laid out.
Will it be fixed in next PRs?

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte [hunk 9]
```diff
@@ -0,0 +1,1153 @@
...
+      loadingIssues = false
+    },
+    {
+      sort: { startDate: SortingOrder.Ascending, rank: SortingOrder.Ascending }
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttView.svelte]: ganttViewOptions() offers three sort options (startDate/rank/dueDate), but here sort is hardcoded as { startDate, rank }

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte [hunk 10]
```diff
@@ -0,0 +1,1153 @@
...
+  import { buildLayout } from './lib/layout'
+  import { createTimeScale } from './lib/time-scale'
+  import { type LayoutRow, type MilestoneMarker, type SummaryRange, type ZoomLevel } from './lib/types'
+  import CreateIssue from '../CreateIssue.svelte'
+  import { statusStore } from '@hcengineering/view-resources'
+  import ArrowLeft from '@hcengineering/ui/src/components/icons/ArrowLeft.svelte'
+  import ArrowRight from '@hcengineering/ui/src/components/icons/ArrowRight.svelte'
+  import NavPrev from '@hcengineering/ui/src/components/icons/NavPrev.svelte'
+  import NavNext from '@hcengineering/ui/src/components/icons/NavNext.svelte'
+  import Calendar from '@hcengineering/ui/src/components/icons/Calendar.svelte'
```
Comment (ArtyomSavchenko) [inline plugins/tracker-resources/src/components/gantt/GanttView.svelte]: These icons are all exported from the package root (IconArrowLeft, IconArrowRight, IconNavPrev, IconNavNext, IconCalendar), I think there is no need to use deep imports from src

File: plugins/tracker-resources/src/components/issues/IssuesView.svelte (hunk 1 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/issues/IssuesView.svelte:13]: `CreateIssue` already accepts `space` as optional. The same undefined-space path is used by `MyIssues`, and `NewIssueHeader` passes an optional `currentSpace` through to `CreateIssue`. The built-in `SpaceSelector` auto-selects a project, and Save stays gated until a project/status/kind resolve — so no guard is needed here.

File: plugins/tracker-resources/src/components/gantt/GanttTodayMarker.svelte (hunk 2 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttTodayMarker.svelte]: Good catch — a real inconsistency. The line sits on the UTC grid, but the label used the local date, so far from UTC the pill could show a different day than its column. Fixed in 3032588 by rendering the Today label with `timeZone: 'UTC'`, matching the month ticks.

File: models/tracker/src/viewlets.ts (hunk 3 above)
Comment (MichaelUray) [reply models/tracker/src/viewlets.ts]: You're right — that was an internal note about the PR series that leaked into the code. Rephrased in 3032588 to describe the behaviour without referencing the PR.

File: plugins/tracker-resources/src/components/gantt/lib/time-scale.ts (hunk 4 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/lib/time-scale.ts:34]: `fromX` is actually already used in this PR (`GanttCanvas` / `GanttHeader` map pointer positions back to dates), and the editable viewlet in #10854 builds on it for drag-to-reschedule — so it isn't test-only.

File: plugins/tracker-resources/src/components/gantt/GanttBar.svelte (hunk 5 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttBar.svelte:33]: Agreed these shouldn't stay hardcoded. The bar-color system is reworked in #10873 to derive from the platform status palette (theme variables + `getPlatformColor`); these are interim defaults for the read-only viewlet. Happy to pull the palette forward into this PR if you'd rather not merge the interim values.

File: plugins/tracker-assets/lang/de.json (hunk 6 above)
Comment (MichaelUray) [reply plugins/tracker-assets/lang/de.json]: Fixed the Gantt locale keys for German and Russian in 3032588. The remaining locale files keep the existing English fallback values for now; I'd rather leave those as fallback strings than add unreviewed machine translations for languages I can't validate. If you'd prefer, I can also seed best-effort translations for the other locales.

File: plugins/tracker-resources/src/components/gantt/GanttCanvas.svelte (hunk 7 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttCanvas.svelte]: Good catch — throttled the hover dispatch to ~one call per frame in 3032588, so it no longer fires on every mousemove pixel. I kept the mouse coordinates in the event rather than dispatching only on row change, since the tooltip is positioned from `mouseX/mouseY` and would otherwise go stale while the cursor moves.

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte (hunk 8 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttView.svelte:125]: Fair point on the query limit. #10858 (tier-3) adds row/render virtualization — only the visible rows are laid out and painted — but you're right the query itself still loads the full filtered set; a query-level limit/pagination would be a separate follow-up rather than something the current series adds. I can add a sane cap here if you'd like.

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte (hunk 9 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttView.svelte]: You're right — the view advertised three sort options (startDate / rank / dueDate) that the viewlet was silently ignoring. Wired the selected `orderBy` from the view options in 3032588 (startDate as the default, rank as a stable tiebreaker), following the `GridView` / `CardGridView` pattern.

File: plugins/tracker-resources/src/components/gantt/GanttView.svelte (hunk 10 above)
Comment (MichaelUray) [reply plugins/tracker-resources/src/components/gantt/GanttView.svelte]: Fixed in 3032588 — switched to the package-root exports (`IconArrowLeft` / `IconArrowRight`, `IconNavPrev` / `IconNavNext`, `IconCalendar`) instead of the deep `src/...` imports.

**Comment by ArtyomSavchenko** [comment]
Hi @MichaelUray 
What do you think about leaving only one pull request that includes all the Gantt changes?
This will make it easier for me to see the whole picture and test it.

**Comment by MichaelUray** [comment]
Hi @ArtyomSavchenko — done. I consolidated #10853–#10859, #10872 and #10873 into a single pull request: https://github.com/hcengineering/platform/pull/10992

It contains the complete Gantt series plus the fixes made during consolidation, rebased onto current develop. Fork CI on the exact head is green, including uitest and uitest-pg.

I am closing the nine superseded pull requests with links to the consolidated PR so review can continue in one place.
